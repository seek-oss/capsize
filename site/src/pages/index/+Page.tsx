import React, { ReactNode } from 'react';
import { Stack, Box, Text, Link } from '@chakra-ui/react';

import { AppStateProvider } from '../../components/AppStateContext';
import { SiteProvider } from '../../components/SiteProvider';
import Seo from '../../components/Seo';
import Layout from '../../components/Layout';
import FontSelector from '../../components/FontSelector';
import OutputCSS from '../../components/OutputCSS';
import Logo from '../../components/Logo';
import CapSizeSelector from '../../components/CapSizeSelector';
import Preview from '../../components/Preview';
import MetricsPreview from '../../components/MetricsPreview';
import ContentBlock from '../../components/ContentBlock';
import Heading from '../../components/Heading';
import FAQs from '../../components/FAQs';
import GitHubStars from '../../components/GitHubStars';
import { px } from '../../utils';

const Step = ({
  number,
  title,
  children,
}: {
  number?: number;
  title?: string;
  children: ReactNode;
}) => (
  <Stack spacing={10}>
    {title && (
      <Box>
        <ContentBlock>
          <Heading as="h2" size="2">
            <Box as="span" color="pink.400" fontSize={['1.2em', '1.5em']}>
              {number}.{' '}
            </Box>
            {title}
          </Heading>
        </ContentBlock>
      </Box>
    )}
    <Box>{children}</Box>
  </Stack>
);

const IndexPage = () => (
  <AppStateProvider>
    <SiteProvider>
      <Layout>
        <Seo />
        <Stack spacing={[10, 20]}>
          <Box paddingTop={[10, 20]}>
            <ContentBlock>
              <Stack spacing={10}>
                <Stack spacing={[12, 12, 16]}>
                  <Box
                    display="flex"
                    flexDirection="column"
                    alignItems="center"
                  >
                    <Box marginBottom={[4, 4, 8]} boxSize={px([80, 100, 120])}>
                      <Logo />
                    </Box>
                    <Heading size="1">Capsize</Heading>
                  </Box>
                  <Box>
                    <Heading as="div" size="3" align="center">
                      <Box whiteSpace="nowrap">
                        Flipping how we define
                        <br />
                        typography in CSS
                      </Box>
                    </Heading>
                  </Box>
                  <Box display="flex" justifyContent="center" paddingX={2}>
                    <Text
                      fontSize={[18, 18, 20]}
                      color="#66748A"
                      maxWidth="48ex"
                      textAlign="center"
                    >
                      Capsize makes the sizing and layout of text as predictable
                      as every other element on the screen.
                      <br />
                      <br />
                      Using font metadata, text can now be sized according to
                      the height of its capital letters while trimming the space
                      above capital letters and below the baseline.
                    </Text>
                  </Box>
                </Stack>
                <Box
                  display="flex"
                  justifyContent="center"
                  flexDirection={['column', 'row']}
                >
                  <Text
                    fontSize={[16, 16, 20]}
                    fontFamily="monospace"
                    textAlign="center"
                    bg="gray.100"
                    paddingY={3}
                    paddingX={6}
                    borderRadius={24}
                    marginBottom={[4, 0]}
                  >
                    npm install @capsizecss/core
                  </Text>
                  <Link
                    href="https://github.com/seek-oss/capsize"
                    display="flex"
                    justifyContent="center"
                    marginLeft={[0, 4]}
                    borderRadius={24}
                    boxShadow="0 0 0 2px #edf2f7"
                    color="#66748A"
                    paddingX={4}
                    paddingY={[2, 0]}
                    _hover={{ background: '#f1f5f9', color: 'inherit' }}
                  >
                    <GitHubStars />
                  </Link>
                </Box>
              </Stack>
            </ContentBlock>
          </Box>

          <Box background="white" paddingY={[10, 20]}>
            <Stack spacing={[10, 10, 10, 20]}>
              <Box>
                <Step number={1} title="Choose a font">
                  <ContentBlock>
                    <Stack spacing={6}>
                      <Box>
                        <FontSelector />
                      </Box>
                      <Box>
                        <MetricsPreview />
                      </Box>
                    </Stack>
                  </ContentBlock>
                </Step>
              </Box>

              <Box>
                <Step number={2} title="Adjust size & spacing">
                  <Stack spacing={10}>
                    <Box>
                      <ContentBlock>
                        <CapSizeSelector />
                      </ContentBlock>
                    </Box>
                    <Box>
                      <Preview />
                    </Box>
                  </Stack>
                </Step>
              </Box>

              <Box>
                <Step number={3} title="Apply the styles">
                  <ContentBlock>
                    <OutputCSS />
                  </ContentBlock>
                </Step>
              </Box>
            </Stack>
          </Box>

          <Box as="section">
            <ContentBlock>
              <FAQs />
            </ContentBlock>
          </Box>
        </Stack>
      </Layout>
    </SiteProvider>
  </AppStateProvider>
);

export { IndexPage as Page };
