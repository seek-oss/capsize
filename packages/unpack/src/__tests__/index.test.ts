import { readFile } from 'node:fs/promises';
import { fromUrl, fromBlob, fromBuffer } from '../index';
import { join } from 'node:path';

describe('unpack font metrics', () => {
  describe('fromBuffer', () => {
    it('should return the font metrics', async () => {
      const fontPath = join(__dirname, './__fixtures__/NotoSans-Regular.ttf');
      const font = await readFile(fontPath);
      const metrics = await fromBuffer(font);
      expect(metrics).toMatchSnapshot();
    });
  });

  describe('fromBlob', () => {
    it('should return the font metrics', async () => {
      const fontPath = join(__dirname, './__fixtures__/NotoSans-Regular.ttf');
      const font = await readFile(fontPath);
      const fontBlob = new Blob([font]);

      const metricsBlob = await fromBlob(fontBlob);
      expect(metricsBlob).toMatchSnapshot();

      const metricsFont = await fromBuffer(font);
      expect(metricsFont).toEqual(metricsBlob);
    });
  });

  describe('fromUrl', () => {
    it('should return the font metrics', async () => {
      const fontPath = join(__dirname, './__fixtures__/NotoSans-Regular.ttf');
      const font = await readFile(fontPath);
      const sampleFontUrl =
        'https://github.com/notofonts/notofonts.github.io/raw/refs/heads/main/fonts/NotoSans/full/ttf/NotoSans-Regular.ttf';

      const metricsUrl = await fromUrl(sampleFontUrl);
      expect(metricsUrl).toMatchSnapshot();

      const metricsFont = await fromBuffer(font);
      expect(metricsFont).toEqual(metricsUrl);
    });
  });
});
