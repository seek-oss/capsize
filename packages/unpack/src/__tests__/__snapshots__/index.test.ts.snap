// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`unpack font metrics fromBlob should return the font metrics 1`] = `
{
  "ascent": 1069,
  "capHeight": 714,
  "descent": -293,
  "familyName": "Noto Sans",
  "fullName": "Noto Sans Regular",
  "lineGap": 0,
  "postscriptName": "NotoSans-Regular",
  "subsets": {
    "latin": {
      "xWidthAvg": 474,
    },
    "thai": {
      "xWidthAvg": 600,
    },
  },
  "unitsPerEm": 1000,
  "xHeight": 536,
  "xWidthAvg": 474,
}
`;

exports[`unpack font metrics fromBuffer should return the font metrics 1`] = `
{
  "ascent": 1069,
  "capHeight": 714,
  "descent": -293,
  "familyName": "Noto Sans",
  "fullName": "Noto Sans Regular",
  "lineGap": 0,
  "postscriptName": "NotoSans-Regular",
  "subsets": {
    "latin": {
      "xWidthAvg": 474,
    },
    "thai": {
      "xWidthAvg": 600,
    },
  },
  "unitsPerEm": 1000,
  "xHeight": 536,
  "xWidthAvg": 474,
}
`;

exports[`unpack font metrics fromUrl should return the font metrics 1`] = `
{
  "ascent": 1069,
  "capHeight": 714,
  "descent": -293,
  "familyName": "Noto Sans",
  "fullName": "Noto Sans Regular",
  "lineGap": 0,
  "postscriptName": "NotoSans-Regular",
  "subsets": {
    "latin": {
      "xWidthAvg": 474,
    },
    "thai": {
      "xWidthAvg": 600,
    },
  },
  "unitsPerEm": 1000,
  "xHeight": 536,
  "xWidthAvg": 474,
}
`;
